### ISSUE TITLE
[gateway-gcs] ListObjects does not work when marker is used

### ISSUE THREAD

**Issue by krishnasrinivas** [minio/minio#4411]
Make this change in `mc`:
```
krishna@escape:~/gocode/src/github.com/minio/mc$ git diff
diff --git a/vendor/github.com/minio/minio-go/api-list.go b/vendor/github.com/minio/minio-go/api-list.go
index a19156a..3924c74 100644
--- a/vendor/github.com/minio/minio-go/api-list.go
+++ b/vendor/github.com/minio/minio-go/api-list.go
@@ -110,7 +110,7 @@ func (c Client) ListObjectsV2(bucketName, objectPrefix string, recursive bool, d
                var continuationToken string
                for {
                        // Get list of objects a maximum of 1000 per request.
-                       result, err := c.listObjectsV2Query(bucketName, objectPrefix, continuationToken, fetchOwner, delimiter, 1000)
+                       result, err := c.listObjectsV2Query(bucketName, objectPrefix, continuationToken, fetchOwner, delimiter, 10)
                        if err != nil {
                                objectStatCh <- ObjectInfo{
                                        Err: err,
@@ -288,7 +288,7 @@ func (c Client) ListObjects(bucketName, objectPrefix string, recursive bool, don
                var marker string
                for {
                        // Get list of objects a maximum of 1000 per request.
-                       result, err := c.listObjectsQuery(bucketName, objectPrefix, marker, delimiter, 1000)
+                       result, err := c.listObjectsQuery(bucketName, objectPrefix, marker, delimiter, 10)
                        if err != nil {
                                objectStatCh <- ObjectInfo{
                                        Err: err,
diff --git a/vendor/github.com/minio/minio-go/constants.go b/vendor/github.com/minio/minio-go/constants.go
index 057c3ee..baccefd 100644
--- a/vendor/github.com/minio/minio-go/constants.go
+++ b/vendor/github.com/minio/minio-go/constants.go
@@ -20,7 +20,7 @@ package minio
 
 // miniPartSize - minimum part size 64MiB per object after which
 // putObject behaves internally as multipart.
-const minPartSize = 1024 * 1024 * 64
+const minPartSize = 1024 * 1024 * 5
 
 // maxPartsCount - maximum number of parts for a single multipart session.
 const maxPartsCount = 10000
krishna@escape:~/gocode/src/github.com/minio/mc$ 
```

create 20 objects under prefix `a/`
```
for i in {1..20}; do mc cp /etc/passwd myminio/krisbuck/a/$i; done
```

Then list gives error:
```
krishna@escape:~$ mc ls myminio/krisbuck/a/
[2017-05-23 12:26:30 PDT] 2.2KiB 1
[2017-05-23 12:26:50 PDT] 2.2KiB 10
[2017-05-23 12:26:52 PDT] 2.2KiB 11
[2017-05-23 12:26:54 PDT] 2.2KiB 12
[2017-05-23 12:26:56 PDT] 2.2KiB 13
[2017-05-23 12:26:58 PDT] 2.2KiB 14
[2017-05-23 12:27:00 PDT] 2.2KiB 15
[2017-05-23 12:27:02 PDT] 2.2KiB 16
[2017-05-23 12:27:03 PDT] 2.2KiB 17
[2017-05-23 12:27:05 PDT] 2.2KiB 18
mc: <ERROR> Unable to list folder. A header you provided implies functionality that is not implemented
krishna@escape:~$
```

`mc --debug`:
```
mc: <DEBUG> GET /krisbuck/?delimiter=%2F&marker=CgRhLzE4&max-keys=10&prefix=a%2F HTTP/1.1
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.4 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170523/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170523T194554Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 501 Not Implemented
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: application/xml
Date: Tue, 23 May 2017 19:45:54 GMT
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Vary: Origin
X-Amz-Bucket-Region: 
X-Amz-Request-Id: 14C153DBB6CB419B

11c
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NotImplemented</Code><Message>A header you provided implies functionality that is not implemented</Message><Key></Key><BucketName></BucketName><Resource>/krisbuck/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

```

validateListObjectsArgs() is returning error:
```
        if marker != "" {
                // Marker not common with prefix is not implemented.
                if !hasPrefix(marker, prefix) {
                        return ErrNotImplemented
                }
        }
```

**Comment by krishnasrinivas** [comment]
Two options:
* disable `hasPrefix(marker, prefix)` check. If client app sets marker (object name) then it won't be valid for backend GCS
* Support only ListObjectsV2 for GCS which has support for continuation token.

Looks like GCS ListObjects was not designed for the user to specify/construct Token, but instead only to use previously returned nextPageToken:

https://cloud.google.com/storage/docs/json_api/v1/objects/list

**Comment by krisis** [comment]
@krishnasrinivas 
> disable hasPrefix(marker, prefix) check. If client app sets marker (object name) then it won't be valid for backend GCS

Do you propose disabling the check only for GCS?

> Support only ListObjectsV2 for GCS which has support for continuation token.

I prefer this option since it aligns with GCS API.

**Comment by krisis** [comment]
A simpler way to reproduce this issue,

1. `mc mb mygcs/mybucket`
2. `mc cp /etc/lsb-release mygcs/mybucket/obj1`
3. `mc cp /etc/lsb-release mygcs/mybucket/obj2`
4.  Run the following code.

```go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	minio "github.com/minio/minio-go"
)

func main() {
	clnt, err := minio.NewCore("localhost:9000", "minio", "minio123", false)
	if err != nil {
		log.Fatalln(err)
	}

	bucket := "mybucket"
	prefix := "o"
	delimiter := ""
	marker := ""
	for {
		res, err := clnt.ListObjects(bucket, prefix, marker, delimiter, 1)
		if err != nil {
			log.Fatalln(err)
		}

		jsonBytes, err := json.MarshalIndent(res, " ", "\t")
		if err != nil {
			log.Fatalln(err)
		}
		fmt.Println(count, string(jsonBytes))
		if !res.IsTruncated {
			break
		}
		marker = res.NextMarker
	}
}

```

**Comment by harshavardhana** [comment]
> Do you propose disabling the check only for GCS

No we can do this for every backend. i.e for our server backends like erasure code and fs we return empty listObjects response rather than an error.  In case of backends which do actually handle it such as GCS and S3 we can send it all the way to the remote site.

**Comment by krisis** [comment]
@krishnasrinivas @harshavardhana I will remove the `hasPrefix` check from `ListObjectsV1` method of gateway handlers. This moves the validation to the respective gateway backends (like azure, gcs, s3). Does this make sense?

**Comment by deekoder** [comment]
Need to discuss.

**Comment by krishnasrinivas** [comment]
for ListV1, in the response we will set marker as the continuation token returned by GCS just like how we do today, but with small change, we'll prefix the token with ##minio
This way if the client sends a marker with this prefix we know that we need to use it as token for GCS List
If the marker does not have ##minio prefix we'll List the GCS from the beginning and wait till we reach the marker so that we can send the client appropriate response


For ListV2, the implementation is more straight forward as both S3 and GCS will operate on a token. However if the client sends `start-after` then we will list from the server to filter out the entries till that entry.

**Comment by krisis** [comment]
@krishnasrinivas The above solution allows S3-compatible applications to use `marker` as they would with Minio server or AWS S3. Applications using non-overlapping `marker` and `prefix` would continue to fail with `NotImplemented` error. We need to remove `hasPrefix` check from `ListObjectsV1Handler` in `gateway-handlers.go` to address this issue completely.

**Comment by krisis** [comment]
@harshavardhana this issue is fixed by https://github.com/minio/minio/pull/4495, can be closed.

**Comment by harshavardhana** [comment]
Fixed in #4495 - closing.
